### ISSUE TITLE
Setup of CnnLossLayer and Softmax - Help request

### ISSUE THREAD

**Issue by bekkon** [deeplearning4j/deeplearning4j#5705]
hi,

I need a bit of an advice again. I am trying to set up a neural net for Image Segmentation and got stuck with the Classifier part. (Started discussion with @AlexDBlack on gitter)

See a simplified net here: https://gist.github.com/bekkon/ac7e4ab4db3fbe2e1b97a0c3e838b19c

I am trying to get a Convolution layer with the Softmax activation function working, but I get a "Mismatched shapes (shape = [2, 2, 256, 256], column vector shape =[2, 256, 256])" exception. I can get the net working if I move the Softmax activation function into the CnnLossLayer, but this is not something I want, as I'd like the classifier layer to take multiple channels and calculate its output accordingly. As I see, if I move the Softmax to the CnnLossLayer, its input Channels will already be reduced according to the nOut parameter of the preceding layer - so it is not an equivalent transformation of the network.

How could I get the Convolutional layer to work with the Softmax? What am I doing wrong?

**Comment by AlexDBlack** [comment]
Thanks for the issue. Now that I've had time to look at your network configuration properly, I can see that the problem is the network configuration.

> As I see, if I move the Softmax to the CnnLossLayer, its input Channels will already be reduced according to the nOut parameter of the preceding layer - so it is not an equivalent transformation of the network.

This is incorrect. I'm not following your reasoning here, but I can say that your conclusion of it not being an equivalent configuration is certainly incorrect.

One restriction of our softmax activation function is that it's basically set up to function on 2d activations. CnnLossLayer accounts for this restriction by - internally only - reshaping to 2d - i.e., the N (N=2 for your case) channels are processed separately for each x/y location in the output. (it reshapes back to 4d later)

tl;dr conv(softmax) + cnnLoss(identity) is definitely equivalent to conv(identity) + cnnLoss(softmax) - except that the latter accounts for our softmax 2d restriction hence actually works.

For segmentation, CnnLossLayer configured with softmax activation function is what you want.

**Comment by bekkon** [comment]
Thank you. Now I get it. I had a few concepts mixed up, hence the wrong reasoning. Reading your answer helped me sort out my misunderstandings about the way convolution / activation / loss work together. Thanks again :)
